python/oneflow/test/graph/test_graph_io_check.py

```diff
@@ -50,17 +50,17 @@ def fn(*args, **kwargs):
             inp = (args, kwargs)
             print("origin: ", inp)
 
-            io_node = IONode(None, 0, inp, "Graph_0")
+            io_node, named_nodes = NamedIONode.construct(inp, "Graph_0", None)
 
-            for (name, node) in list(io_node.named_nodes()):
+            for (name, node) in named_nodes:
                 print(name, repr(node))
 
             def leaf_fn(node):
-                if isinstance(node._value, str):
+                if isinstance(node.value(), str):
                     return "mapped_str"
-                return node._value
+                return node.value()
 
-            m_v = io_node.map_leaf(leaf_fn)
+            m_v = map_structed_value_leaf(io_node, leaf_fn)
```

Comment:
Can not see any benifit of refactoring OOP style to function style.
- inconsistency of interface, a class construct and an outside map function;
- an outside map function is not high Cohesion and low Coupling;
- default to return a flat list is not good and ugly;

Reply:
The IONode before refactoring consists of the following two functionalities:
1. associate each input/output argument with a prefix, name and index.
2. map the leaf IONode given the map function.

1 and 2 are coupled together in the old design. If you want to use the map function, you must turn the input/output arguments into IONode first. 

After refactoring, you can map arbitrary input/output arguments without turning them into IONode if not necessary.
For example, we have the following input arguments.

`args = [tensor1, "test", {"i": tensor3}] `

if we want to map each **tensor** in args into the contiguous tensor, we can do it as the following:

```
def map_func(value):
     if isinstance(value, Tensor):
            value.contiguous_()
     return value

map_strcuted_value_leaf(args, map_func)
```
In the above case, convert the args into IONode is not needed. 

===============================================

The `NamedIONode.construct ` returning an additional flattened list is for efficiency reason, because you don't have to recursively traverse the IONode again to get the flattened list.  @strint

Reply:
What about An IONode with a  bool called `gen_name` to fit both needs:
- same API 
  - map_leaf api for  map
  - an iter api  for iteration operation
  - an flatten API to return the flattened (cached) result
  - for efficiency, change tree build to lazy evaluate, trigger tree build when map or flat is called